src/OpenCensus/Trace/Propagation/TraceContextFormat.cs

```diff
@@ -0,0 +1,136 @@
+
+        /// <inheritdoc/>
+        public override ISpanContext Extract<T>(T carrier, Func<T, string, IEnumerable<string>> getter)
+        {
+            // from https://github.com/w3c/distributed-tracing/blob/master/trace_context/HTTP_HEADER_FORMAT.md
+            // traceparent: 00-0af7651916cd43dd8448eb211c80319c-00f067aa0ba902b7-01
+
+            var traceparent = getter(carrier, "traceparent")?.FirstOrDefault();
+            var tracestateCollection = getter(carrier, "tracestate");
+
+            if (traceparent == null)
+            {
+                return null;
+            }
+
+            // TODO: validate version prefix
+            var traceId = TraceId.FromBytes(Arrays.StringToByteArray(traceparent, "00-".Length, "0af7651916cd43dd8448eb211c80319c".Length));
+
+            // TODO: validate span delimeter
+            var spanId = SpanId.FromBytes(Arrays.StringToByteArray(traceparent, "00-0af7651916cd43dd8448eb211c80319c-".Length, "00f067aa0ba902b7".Length));
+
+            // TODO: validate options delimeter
+            var optionsArray = Arrays.StringToByteArray(traceparent, "00-0af7651916cd43dd8448eb211c80319c-00f067aa0ba902b7-".Length, 2);
+
+            var options = TraceOptions.Default;
+            if ((optionsArray[0] | 1) == 1)
+            {
+                options = TraceOptions.Builder().SetIsSampled(true).Build();
+            }
+
+            List<KeyValuePair<string, string>> entries = new List<KeyValuePair<string, string>>();
+            if (tracestateCollection != null)
+            {
+                foreach (var tracestate in tracestateCollection)
+                {
+                    // tracestate: rojo=00-0af7651916cd43dd8448eb211c80319c-00f067aa0ba902b7-01,congo=BleGNlZWRzIHRohbCBwbGVhc3VyZS4
+                    var keyStartIdx = 0;
+                    var length = tracestate.Length;
+                    while (keyStartIdx < length)
+                    {
+                        var keyEndIdx = tracestate.IndexOf("=", keyStartIdx);
+
+                        if (keyEndIdx == -1)
+                        {
+                            // error happened. Ignore this tracestate
+                            break;
+                        }
+
+                        var valueStartIdx = keyEndIdx + 1;
+
+                        var valueEndIdx = tracestate.IndexOf(",", valueStartIdx);
+                        valueEndIdx = valueEndIdx == -1 ? length : valueEndIdx;
+                        entries.Add(
+                            new KeyValuePair<string, string>(
+                                tracestate.Substring(keyStartIdx, keyEndIdx - keyStartIdx).Trim(), 
+                                tracestate.Substring(valueStartIdx, valueEndIdx - valueStartIdx).Trim()));
+                        keyStartIdx = valueEndIdx + 1;
+                    }
+                }
+            }
+
+            var tracestateBuilder = Tracestate.Builder;
+
+            entries.Reverse();
```

Comment:
can we optimize creating an extra array and reversing it?

Reply:
maybe another API on Tracestate... or parsing backwards... Do you have a suggestion? My understanding is that reversing is not very expensive operation. Alternatively I can iterate backwards